src/snowflake/snowpark/dataframe_reader.py

```diff
@@ -1285,6 +1298,156 @@ def _read_semi_structured_file(self, path: str, format: str) -> DataFrame:
         return df
 
+    @private_preview(version="1.38.0")
+    @publicapi
+    def jdbc(
+        self,
+        url: str,
+        udtf_configs: dict,
+        *,
+        properties: Optional[dict] = None,
+        table: Optional[str] = None,
+        query: Optional[str] = None,
+        column: Optional[str] = None,
+        lower_bound: Optional[Union[str, int]] = None,
+        upper_bound: Optional[Union[str, int]] = None,
+        num_partitions: Optional[int] = None,
+        query_timeout: Optional[int] = 0,
+        fetch_size: Optional[int] = 0,
+        custom_schema: Optional[Union[str, StructType]] = None,
+        predicates: Optional[List[str]] = None,
+        session_init_statement: Optional[List[str]] = None,
+        _emit_ast: bool = True,
+    ) -> DataFrame:
+        """
+        Reads data from a database table or query into a DataFrame using a DBAPI connection,
+        with support for optional partitioning, parallel processing, and query customization.
+
+        There are multiple methods to partition data and accelerate ingestion.
+        These methods can be combined to achieve optimal performance:
+
+        1.Use column, lower_bound, upper_bound and num_partitions at the same time when you need to split large tables into smaller partitions for parallel processing.
+        These must all be specified together, otherwise error will be raised.
+        2.Set max_workers to a proper positive integer.
+        This defines the maximum number of processes and threads used for parallel execution.
+        3.Adjusting fetch_size can optimize performance by reducing the number of round trips to the database.
+        4.Use predicates to defining WHERE conditions for partitions,
+        predicates will be ignored if column is specified to generate partition.
+        5.Set custom_schema to avoid snowpark infer schema, custom_schema must have a matched
+        column name with table in external data source.
+
+        Args:
+            url: A connection string used to establish connections to external data source with JDBC driver.
+                Please do not include any secrets in this parameter.
+            udtf_configs: A dictionary containing configuration parameters for ingesting external data using a Snowflake UDTF.
+                This parameter is required for jdbc.
+
+                The dictionary may include the following keys:
+
+                - external_access_integration (str, required): The name of the external access integration,
+                    which allows the UDTF to access external endpoints.
+
+                - imports (List[str], required): A list of stage file names to import into the UDTF.
+                    Please include Jar file of jdbc driver to establish connection to external data source.
+
+                - java_version (int, optional): A integer that indicate the java runtime version of udtf.
+                    By default we use java 11.
```

Comment:
emmm why not default to 17? usually the newer version is faster than the older one.
btw is our own JDBC code compatible with both 11 and 17?